File: algorithmic_transit_simulator/transit_simulator/disjoint_set.py
class DisjointSet:
    """Union-find with path compression and union by size."""
    def __init__(self, items=()):
        self.parent = {item: item for item in items}; self.size = {item: 1 for item in items}

    def add(self, item):
        if item not in self.parent: self.parent[item] = item; self.size[item] = 1

    def find(self, item):
        self.add(item)
        if self.parent[item] != item: self.parent[item] = self.find(self.parent[item])
        return self.parent[item]

    def union(self, left, right):
        left, right = self.find(left), self.find(right)
        if left == right: return False
        if self.size[left] < self.size[right]: left, right = right, left
        self.parent[right] = left; self.size[left] += self.size[right]
        return True

    def connected(self, left, right): return self.find(left) == self.find(right)

File: structures/disjoint_set.go
package structures

// DisjointSet tracks connected components using union by size and path compression.
type DisjointSet struct{ parent,size []int }
func NewDisjointSet(n int)*DisjointSet{p:=make([]int,n);s:=make([]int,n);for i:=range p{p[i]=i;s[i]=1};return &DisjointSet{p,s}}
func(d *DisjointSet) Find(x int)int{if d.parent[x]!=x{d.parent[x]=d.Find(d.parent[x])};return d.parent[x]}
func(d *DisjointSet) Union(a,b int)bool{a,b=d.Find(a),d.Find(b);if a==b{return false};if d.size[a]<d.size[b]{a,b=b,a};d.parent[b]=a;d.size[a]+=d.size[b];return true}
func(d *DisjointSet) Connected(a,b int)bool{return d.Find(a)==d.Find(b)}
